File: src/main/python/AoC2016_16.py
#! /usr/bin/env python3
#
# Advent of Code 2016 Day 16
#

import sys
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase


class DragonCurve(NamedTuple):
    initial_state: str
    table: dict[int, int] = str.maketrans("01", "10")

    def dragon_curve(self, input_: str) -> str:
        return input_ + "0" + input_[::-1].translate(self.table)

    def checksum(self, data: list[str]) -> str:
        pairs = [
            "1" if data[i] == data[i + 1] else "0"
            for i in range(0, len(data) - 1, 2)
        ]
        return "".join(pairs) if len(pairs) % 2 != 0 else self.checksum(pairs)

    def checksum_for_size(self, size: int) -> str:
        data = self.initial_state
        while len(data) < size:
            data = self.dragon_curve(data)
        return self.checksum(list(data[:size]))


Input = DragonCurve
Output1 = str
Output2 = str


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return DragonCurve(list(input_data)[0])

    def solve(self, dragon_curve: DragonCurve, size: int) -> str:
        return dragon_curve.checksum_for_size(size)

    def part_1(self, dragon_curve: DragonCurve) -> str:
        return self.solve(dragon_curve, 272)

    def part_2(self, dragon_curve: DragonCurve) -> str:
        return self.solve(dragon_curve, 35651584)

    def samples(self) -> None:
        assert self.solve(self.parse_input(["10000"]), 20) == "01100"


solution = Solution(2016, 16)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2016_16.java
import java.util.List;

import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringUtils;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2016_16
        extends SolutionBase<AoC2016_16.DragonCurve, String, String> {
	
	private AoC2016_16(final boolean debug) {
		super(debug);
	}
	
	public static AoC2016_16 create() {
		return new AoC2016_16(false);
	}

	public static AoC2016_16 createDebug() {
		return new AoC2016_16(true);
	}
	
	private String solve(final DragonCurve dragonCurve, final int size) {
	    return dragonCurve.checkSumForSize(size);
	}

	@Override
    protected DragonCurve parseInput(final List<String> inputs) {
        return new DragonCurve(inputs.get(0));
    }

    @Override
	public String solvePart1(final DragonCurve dragonCurve) {
	    return solve(dragonCurve, 272);
	}
	
	@Override
	public String solvePart2(final DragonCurve dragonCurve) {
	    return solve(dragonCurve, 35651584);
	}
	
	@Override
    public void samples() {
	    final AoC2016_16 test = AoC2016_16.createDebug();
	    assert test.solve(test.parseInput(List.of(TEST)), 20).equals("01100");
    }

    public static void main(final String[] args) throws Exception {
        AoC2016_16.create().run();
	}
	
	private static final String TEST = "10000";
	
	record DragonCurve(String initialState) {
	    private String dragonCurve(final String input) {
	        final String b = StringOps.translate(
	                StringUtils.reverse(input), "01", "10");
	        return new StringBuilder().append(input)
	                .append("0").append(b).toString();
	    }
	    
	    private char[] checkSum(final char[] data) {
	        final char[] pairs = new char[data.length / 2];
	        for (int i = 0; i < data.length - 1; i += 2) {
	            if (data[i] == data[i + 1]) {
	                pairs[i / 2] = '1';
	            } else {
	                pairs[i / 2] = '0';
	            }
	        }
	        if (pairs.length % 2 != 0) {
	            return pairs;
	        } else {
	            return checkSum(pairs);
	        }
	    }
        
	    public String checkSumForSize(final int size) {
            String data = this.initialState;
            while (data.length() < size) {
                data = dragonCurve(data);
            }
            return String.valueOf(checkSum(data.substring(0, size).toCharArray()));
        }
	}
}